### PR TITLE
nrf: Enable source line on tracebacks. Improve run-tests.

### PR DESCRIPTION
**Pull request by mzdaniel**
Improve traceback debugging on nrf targets with enough flash, adding missing source line info and improve run-tests to properly handle opt_level and opt_level_lineno tests.

It will be nice to add line number on tracebacks for nrf targets. This is an important REPL and debugging feature, specially if the board has enough room to acommodate it. This feature adds 204 extra bytes to the firmware. AFAICT, this is a standard feature for all targets with flash >= 1M throught the entire repository and for most targets with exception of minimal, bare-arm and pic16bit.

Currently, run-tests fails on opt_level_lineno for nrf52840 and other nrf targets. This PR adds a feature check to fix the issue.

@glenrub: Other features on the nrf family were also checked, UBINASCII (772 bytes), SYS_STDFILES (268 bytes) and ARRAY_SLICE_ASSIGN (176 bytes) before defining MICROPY_ENABLE_SOURCE_LINE on nrf52832 and conditionally defining it on nrf51822. We certainly can revert the changes on nrf52832 and nrf51822 if 204 extra bytes is too much for them.

Here are the tests performed

> nrf/mpconfigport.h (as on master):
>   #define MICROPY_ENABLE_SOURCE_LINE         (0)
> 
> make BOARD=particle_xenon 
> 
> arm-none-eabi-size build-particle_xenon/firmware.elf
>    text    data     bss     dec     hex filename
>  153308      48    4552  157908   268d4 build-particle_xenon/firmware.elf
> 
> MPY: soft reboot
> MicroPython v1.12-700-g0e6ef4035 on 2020-09-01; XENON with NRF52840
> Type "help()" for more information.
> 
> cd tests; ./run-tests --target nrf --device /dev/ttyACM0
> 
> 534 tests performed (14756 individual testcases)
> 533 tests passed
> 226 tests skipped
> 1 tests failed: opt_level_lineno

--------------------------

> nrf/mpconfigdevice_nrf52840.h  (with this PR):
>   #define MICROPY_ENABLE_SOURCE_LINE         (1)
> 
> make BOARD=particle_xenon
> 
> arm-none-eabi-size build-particle_xenon/firmware.elf
>    text    data     bss     dec     hex filename
>  153512      48    4552  158112   269a0 build-particle_xenon/firmware.elf
> 
> MPY: soft reboot
> MicroPython v1.12-701-gae8394d81 on 2020-09-01; XENON with NRF52840
> Type "help()" for more information.
> 
> cd tests; ./run-tests --target nrf --device /dev/ttyACM0
> 
> 535 tests performed (14760 individual testcases)
> 535 tests passed
> 225 tests skipped
>

### CODE REVIEW

**Comment by mzdaniel** [comment]
Rebased on top of master as 2021-08-05

File: ports/nrf/mpconfigdevice_nrf51822.h [hunk 1]
```diff
@@ -59,3 +59,11 @@
 #define MICROPY_PY_UBINASCII               (1)
 #endif
 #endif
+
+#ifndef MICROPY_ENABLE_SOURCE_LINE
```
Comment (dpgeorge) [inline ports/nrf/mpconfigdevice_nrf51822.h:63]: It's good to keep these config options in the same order as what appears in `py/mpconfig.h`, in which case this one should go before `MICROPY_PY_ARRAY_SLICE_ASSIGN` (and after `MICROPY_VFS`).  Same goes for all additions in this PR.

File: tests/feature_check/lineno.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import sys
```
Comment (dpgeorge) [inline tests/feature_check/lineno.py:1]: I can't see this file being used by `tests/run-tests.py`, I don't think it does anything having it here.

**Review by dpgeorge — CHANGES_REQUESTED** [review]
Thanks for this, and sorry it got lost in the PR queue.  It's a good change, definitely very useful to have line numbers in a traceback.

File: ports/nrf/mpconfigdevice_nrf51822.h (hunk 1 above)
Comment (mzdaniel) [reply ports/nrf/mpconfigdevice_nrf51822.h]: Sure, good to know. Done

File: tests/feature_check/lineno.py (hunk 2 above)
Comment (mzdaniel) [reply tests/feature_check/lineno.py:1]: I think I wasnt sure about how do we want to add new user facing tests to the project. I included it for completeness

Comment (dpgeorge) [reply tests/feature_check/lineno.py:1]: There are 2 ways to skip tests:

Comment (mzdaniel) [reply tests/feature_check/lineno.py]: Appreciate your comments and took note. I'm taking the test out for now and the PR looks in good shape.

**Comment by dpgeorge** [comment]
Thanks for updating.  Rebased and merged in 8645b7c23b9d55ddc4e5927f52de5aa3da05a915

**Pull request closed without merging**